### ISSUE TITLE
Strings not handled consistently

### ISSUE THREAD

**Issue by conartist6** [iter-tools/iter-tools#346]
Right now some methods, when receiving string inputs, give string outputs, however this is not true of all methods. An example of a method which does not do this but could is `concat`, e.g.
```js
concat('a', 'b', 'c'); // Iterable['a', 'b', 'c']
// as opposed to
concat('a', 'b', 'c'); // 'abc'
```

I had previously thought that we wanted to do this generally, which is to say make a method like concat return 'abc'. The more I thought about it though the less I agree with this. The problem is what to do when all the inputs are null. Do you return `Iterable[]` or `''`? There's no way to be sure if the nulls would have been strings or normal iterables, and these are hardly interchangable: `''` is falsy, and `Iterable[]` can't be passed to a method expecting a string.

Additionally you have issues when some inputs are strings and some are iterables. What did the user mean?

For these reasons I think the solution is to require the user to be explicit about what they mean. Methods should *always* return non-string iterables, even if all the inputs are strings.

We should then create a method I'll call `stringFrom` (or `str` for short) which is the equivalent of the current `joinAsString` method. I'd also propose a shorthand for `map(str)` called `stringsFrom` (or `strs` for short), suitable for use on the result of one of the `split` methods.

Now, there is a downside to this approach, which is that strings don't really like to be treated as iterables. The best example is building up a string, as concat does. Here is what an implementation of concat for strings looks like:
```js
export function stringConcat(...strs) {
  let result = '';
  for (const str of strs) {
    if (str != null) {
      result += str;
    }
  }
  return result;
}
```
Once upon a time if you'd read perf advice, it would have told you that this is slow and not such a good idea because it allocates more strings than it should, but this is no longer true because engines have multiple subclasses that implement string, including one that acts like Java's `StringBuilder` which would be used internally in the above code. That builder will be more performant if it receives the original constituent strings in full as opposed to receiving characters from the strings one at a time.

Thus I think we want to keep our string optimizations of methods, but we should only trigger their evaluation if a version of the `str` method is being called on them. I think this is safe because it's just an optimization. If the optimization fails to trigger the result would still be the same, it would just be built up character by character instead of string by string.

**Comment by conartist6** [comment]
So I have to ask myself if it is really desirable to perform the SliceString optimization. Some conversation on the topic can be found [here](https://mrale.ph/blog/2016/11/23/making-less-dart-faster.html).

This is a pretty deep topic though, because iter-tools is broadly is subject to this kind of problem. You don't even need strings to see it. Any time you use `slice` or a similar method, you're holding a reference to what could be a very large source iterable just so that you maintain access to a small part of it. If you don't realize you've done this, you may be in for a surprise down the line.

In either case if you need to avoid such a leak you must reallocate. For an iterable this would mean using `arrayFrom`, for a string I think it would mean using an *unoptimized* `stringFrom` method which would rebuild the result from individual characters. It is important to note that my current implementation of `stringFrom` would not qualify because if the input is already a string, it simply returns it.

**Comment by conartist6** [comment]
Another note, what v8 calls SlicedString, SpiderMonkey calls a "dependent string" or "dependent buffer". So the major engines agree on this optimization. I guess that makes me think that we should have code that facilitates its usage, as I have been planning.

The one thing I think I must do differently is to ensure that the the `stringFrom` method always builds a result up from the source character by character. This way optimizations are relevant internal to a series of chained methods (intended to be iter-tools' strength anyway), yet at the end of a series of chained ops you have a real primitive string that won't be leaking anything.

**Comment by conartist6** [comment]
I'm backtracking on the plans I had to leverage the native `SliceString` optimization in iter-tools. I'm generally backing away from any kind of optimization that may suddenly stop working leading to unpredictable performance. I also don't think it's necessary to drop down to the level of natively slicing strings because the whole point of the browser's `SlicedString` optimization is to permit O(1) implementations of operations like concatenation and splitting. iter-tools methods already ensure that only the required work is done -- in essence we're do a userland implementation of the same kinds of optimizations the browser uses under the hood.

For the reasons discussed above we still want our methods to always return iterables, which must then be built back into strings if that is what they should be.

**Comment by conartist6** [comment]
Resuming work on ripping out the bad stuff I did.

**Comment by conartist6** [comment]
Closed by #377.
